File: app/src/main/java/com/bscan/model/graph/Entity.kt
package com.bscan.model.graph

import java.time.LocalDateTime
import java.util.UUID

/**
 * Base entity class for the graph data model.
 * All trackable items, components, locations, people, etc. extend from this.
 */
abstract class Entity(
    open val id: String = generateId(),
    open val entityType: String,
    open val label: String,
    open val properties: MutableMap<String, PropertyValue> = mutableMapOf(),
    open val metadata: EntityMetadata = EntityMetadata()
) {
    
    companion object {
        fun generateId(): String = UUID.randomUUID().toString()
    }
    
    /**
     * Get a property value with type safety
     */
    inline fun <reified T> getProperty(key: String): T? {
        return properties[key]?.getValue<T>()
    }
    
    /**
     * Set a property value
     */
    fun <T> setProperty(key: String, value: T) {
        properties[key] = PropertyValue.create(value)
        metadata.lastModified = LocalDateTime.now()
    }
    
    /**
     * Check if entity has a property
     */
    fun hasProperty(key: String): Boolean = properties.containsKey(key)
    
    /**
     * Remove a property
     */
    fun removeProperty(key: String) {
        properties.remove(key)
        metadata.lastModified = LocalDateTime.now()
    }
    
    /**
     * Get all property keys
     */
    fun getPropertyKeys(): Set<String> = properties.keys
    
    /**
     * Create a copy of this entity with new ID (for versioning/branching)
     */
    abstract fun copy(newId: String = generateId()): Entity
    
    /**
     * Validate this entity's properties and constraints
     */
    open fun validate(): ValidationResult = ValidationResult.valid()
    
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Entity) return false
        return id == other.id
    }
    
    override fun hashCode(): Int = id.hashCode()
    
    override fun toString(): String = "$entityType($id): $label"
}

/**
 * Metadata about entity lifecycle and management
 */
data class EntityMetadata(
    val created: LocalDateTime = LocalDateTime.now(),
    var lastModified: LocalDateTime = LocalDateTime.now(),
    val version: Int = 1,
    val tags: MutableSet<String> = mutableSetOf(),
    val source: String? = null,  // Where this entity came from (scan, import, user, etc.)
    val confidence: Float = 1.0f  // Confidence level for inferred data
)

/**
 * Result of entity validation
 */
sealed class ValidationResult {
    object Valid : ValidationResult()
    data class Invalid(val errors: List<String>) : ValidationResult()
    
    companion object {
        fun valid() = Valid
        fun invalid(vararg errors: String) = Invalid(errors.toList())
    }
    
    val isValid: Boolean get() = this is Valid
}

File: src/models/entities/Entity.ts
/**
 * Base entity class for the graph data model.
 * All trackable items, components, locations, people, etc. extend from this.
 */

import { v4 as uuidv4 } from 'uuid';

export interface EntityMetadata {
  created: Date;
  lastModified: Date;
  version: number;
  tags: Set<string>;
  source?: string; // Where this entity came from (scan, import, user, etc.)
  confidence: number; // Confidence level for inferred data (0.0 to 1.0)
}

/**
 * Result of entity validation
 */
export abstract class ValidationResult {
  abstract readonly isValid: boolean;
  
  static valid(): ValidationResult.Valid {
    return new ValidationResult.Valid();
  }
  
  static invalid(...errors: string[]): ValidationResult.Invalid {
    return new ValidationResult.Invalid(errors);
  }
}

export namespace ValidationResult {
  export class Valid extends ValidationResult {
    readonly isValid = true;
  }
  
  export class Invalid extends ValidationResult {
    readonly isValid = false;
    
    constructor(public readonly errors: string[]) {
      super();
    }
  }
}

/**
 * Generate unique identifier for entities
 */
export function generateId(): string {
  return uuidv4();
}

/**
 * Base entity class
 */
export abstract class Entity {
  public readonly id: string;
  public readonly entityType: string;
  public readonly label: string;
  public readonly properties: Map<string, PropertyValue>;
  public readonly metadata: EntityMetadata;

  constructor(
    id: string = generateId(),
    entityType: string,
    label: string,
    properties: Map<string, PropertyValue> = new Map(),
    metadata?: Partial<EntityMetadata>
  ) {
    this.id = id;
    this.entityType = entityType;
    this.label = label;
    this.properties = properties;
    this.metadata = {
      created: new Date(),
      lastModified: new Date(),
      version: 1,
      tags: new Set(),
      confidence: 1.0,
      ...metadata
    };
  }

  /**
   * Get a property value with type safety
   */
  getProperty<T = any>(key: string): T | undefined {
    const propertyValue = this.properties.get(key);
    return propertyValue?.getValue<T>();
  }

  /**
   * Set a property value
   */
  setProperty<T>(key: string, value: T): void {
    this.properties.set(key, PropertyValue.create(value));
    this.metadata.lastModified = new Date();
  }

  /**
   * Check if entity has a property
   */
  hasProperty(key: string): boolean {
    return this.properties.has(key);
  }

  /**
   * Remove a property
   */
  removeProperty(key: string): void {
    this.properties.delete(key);
    this.metadata.lastModified = new Date();
  }

  /**
   * Get all property keys
   */
  getPropertyKeys(): string[] {
    return Array.from(this.properties.keys());
  }

  /**
   * Create a copy of this entity with new ID (for versioning/branching)
   */
  abstract copy(newId?: string): Entity;

  /**
   * Validate this entity's properties and constraints
   */
  validate(): ValidationResult {
    return ValidationResult.valid();
  }

  equals(other: any): boolean {
    if (this === other) return true;
    if (!(other instanceof Entity)) return false;
    return this.id === other.id;
  }

  toString(): string {
    return `${this.entityType}(${this.id}): ${this.label}`;
  }
}

/**
 * Import PropertyValue after defining Entity to avoid circular dependency
 */
import { PropertyValue } from './PropertyValue';